File: src/sentinel/scoring.py
"""Signal aggregation.

The combination rule has to satisfy three constraints that pull against each
other:

1. The score must never exceed 1.0.
2. Several weak signals must not outrank one strong signal.
3. More corroborating evidence must still raise the score.

Naive summation violates (1) and (2); taking the maximum violates (3). The rule
used here is **noisy-OR**, the probability that at least one independent signal
is a true positive:

    score = 1 - product(1 - w for w in weights)

It is bounded above by 1.0 by construction, it is monotonically increasing in
the number of signals, and it degrades gracefully: three 0.1 signals combine to
0.271, still well below a single 0.9 signal.
"""

from sentinel.types import Signal


class ScoringEngine:
    """Combines detector signals into a single suspicion score."""

    def __init__(self, threshold: float = 0.5) -> None:
        if not 0.0 <= threshold <= 1.0:
            raise ValueError(f"threshold must be in [0.0, 1.0], got {threshold}")
        self.threshold = threshold

    def aggregate(self, signals: list[Signal]) -> float:
        """Combine signal weights via noisy-OR into a score in [0.0, 1.0]."""
        remaining = 1.0
        for signal in signals:
            weight = min(max(signal.weight, 0.0), 1.0)
            remaining *= 1.0 - weight
        score = 1.0 - remaining
        # Guard against float drift so the contract [0.0, 1.0] holds exactly.
        return min(max(score, 0.0), 1.0)

    def is_blocked(self, score: float) -> bool:
        """True when the score meets or exceeds the configured threshold."""
        return score >= self.threshold

File: web/src/lib/sentinel/scoring.ts
/**
 * Signal aggregation. Mirrors `src/sentinel/scoring.py`.
 *
 * The combination rule must satisfy three constraints that pull against each
 * other:
 *
 *   1. The score never exceeds 1.
 *   2. Several weak signals must not outrank one strong signal.
 *   3. More corroborating evidence must still raise the score.
 *
 * Summation breaks (1) and (2); `Math.max` breaks (3). Noisy-OR satisfies all
 * three: the probability that at least one independent signal is a true
 * positive.
 *
 *   score = 1 - Π(1 - w)
 */

import type { Signal } from "./types";

const clamp01 = (n: number): number => Math.min(Math.max(n, 0), 1);

export class ScoringEngine {
  readonly threshold: number;

  constructor(threshold = 0.5) {
    if (!Number.isFinite(threshold) || threshold < 0 || threshold > 1) {
      throw new RangeError(`threshold must be in [0, 1], got ${threshold}`);
    }
    this.threshold = threshold;
  }

  /** Combine signal weights via noisy-OR into a score in [0, 1]. */
  aggregate(signals: readonly Signal[]): number {
    let remaining = 1;
    for (const signal of signals) {
      remaining *= 1 - clamp01(signal.weight);
    }
    // Clamp again to absorb floating-point drift so the contract holds exactly.
    return clamp01(1 - remaining);
  }

  /** True when the score meets or exceeds the configured threshold. */
  isBlocked(score: number): boolean {
    return score >= this.threshold;
  }
}
